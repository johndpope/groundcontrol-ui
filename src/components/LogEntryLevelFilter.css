.LogEntryLevelFilter .ui.radio.checkbox {
  margin-right: 2em;
}
