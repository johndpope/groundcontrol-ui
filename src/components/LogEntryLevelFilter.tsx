// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { Component } from "react";
import { Radio } from "semantic-ui-react";

import "./LogEntryLevelFilter.css";

export interface IProps {
  level: string[] | undefined;
  onChange: (values: IProps) => any;
}

export const allLevel = ["DEBUG", "INFO", "WARNING", "ERROR"];

// Note: we consider undefined level to be the same as all level.
export default class LogEntryLevelFilter extends Component<IProps> {
  public render() {
    const { level } = this.props;
    const radios = allLevel.map((value, i) => (
      <Radio
        key={i}
        label={value}
        checked={!level || level.indexOf(value) >= 0}
        onClick={this.handleToggle.bind(this, value)}
      />
    ));

    return <div className="LogEntryLevelFilter">{radios}</div>;
  }

  private handleToggle(value: string) {
    const level = this.props.level
      ? this.props.level.slice()
      : allLevel.slice();
    const index = level.indexOf(value);

    if (index >= 0) {
      level.splice(index, 1);
    } else {
      level.push(value);
    }

    this.props.onChange({ ...this.props, level });
  }
}
