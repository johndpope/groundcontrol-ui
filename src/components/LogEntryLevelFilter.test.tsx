// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import LogEntryLevelFilter, { allLevel } from "./LogEntryLevelFilter";

const props = {
  level: ["INFO", "WARNING"],
  onChange: jest.fn(),
};

beforeEach(() => {
  mocked(props.onChange).mockClear();
});

describe("<LogEntryLevelFilter />", () => {

  it("renders correctly", () => {
    const wrapper = shallow(<LogEntryLevelFilter {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("checks all the radio buttons when level isn't set", () => {
    const wrapper = shallow(<LogEntryLevelFilter {...props} level={undefined} />);
    expect(wrapper.find("[checked=true]")).toHaveLength(allLevel.length);
  });

  it("triggers onChange when a radio button is unchecked", () => {
    const wrapper = shallow(<LogEntryLevelFilter {...props} />);
    wrapper.find("Radio[label='INFO']").simulate("click");
    expect(props.onChange).toBeCalledTimes(1);
    expect(props.onChange).toBeCalledWith({
      ...props,
      level: ["WARNING"],
    });
  });

  it("triggers onChange when a radio button is checked", () => {
    const wrapper = shallow(<LogEntryLevelFilter {...props} />);
    wrapper.find("Radio[label='DEBUG']").simulate("click");
    expect(props.onChange).toBeCalledTimes(1);
    expect(props.onChange).toBeCalledWith({
      ...props,
      level: ["INFO", "WARNING", "DEBUG"],
    });
  });

});
