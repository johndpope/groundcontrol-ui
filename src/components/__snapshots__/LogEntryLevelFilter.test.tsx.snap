// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogEntryLevelFilter /> renders correctly 1`] = `
<div
  className="LogEntryLevelFilter"
>
  <Radio
    checked={false}
    key="0"
    label="DEBUG"
    onClick={[Function]}
    type="radio"
  />
  <Radio
    checked={true}
    key="1"
    label="INFO"
    onClick={[Function]}
    type="radio"
  />
  <Radio
    checked={true}
    key="2"
    label="WARNING"
    onClick={[Function]}
    type="radio"
  />
  <Radio
    checked={false}
    key="3"
    label="ERROR"
    onClick={[Function]}
    type="radio"
  />
</div>
`;
